conbench/entities/history.py

```diff
@@ -85,31 +89,170 @@ def get_history(case_id, context_id, hardware_hash, repo) -> List[tuple]:
             BenchmarkResult.unit,
             BenchmarkResult.data,
             BenchmarkResult.times,
-            BenchmarkResult.change_annotations,
+            # turn Nones into Falses for this column
+            func.coalesce(
+                BenchmarkResult.change_annotations[
+                    "begins_distribution_change"
+                ].as_boolean(),
+                False,
+            ).label("begins_distribution_change"),
             Hardware.hash,
             Commit.sha,
             Commit.repository,
             Commit.message,
             Commit.timestamp,
-            Distribution.mean_mean,
-            Distribution.mean_sd,
             Run.name,
         )
         .join(Run, Run.id == BenchmarkResult.run_id)
         .join(Hardware, Hardware.id == Run.hardware_id)
         .join(Commit, Commit.id == Run.commit_id)
-        .join(Distribution, Distribution.commit_id == Commit.id)
         .filter(
             BenchmarkResult.case_id == case_id,
             BenchmarkResult.context_id == context_id,
             BenchmarkResult.error.is_(None),
             Commit.sha == Commit.fork_point_sha,  # on default branch
             Commit.repository == repo,
             Hardware.hash == hardware_hash,
-            Distribution.case_id == case_id,
-            Distribution.context_id == context_id,
-            Distribution.hardware_hash == hardware_hash,
         )
-        .order_by(Commit.timestamp.asc())
-        .all()
+        .subquery()
     )
+
+    # Add column with cumulative sum of distribution changes
+    history = Session.query(
+        history,
+        func.sum(history.c.begins_distribution_change.cast(s.Integer))
+        .over(order_by=history.c.timestamp)
+        .label("segment_id"),
+    ).subquery()
+
+    # Add column with rolling mean of the means (only inside of the segment)
+    history = Session.query(
+        history,
+        func.avg(history.c.mean)
+        .over(
+            partition_by=history.c.segment_id,
+            order_by=history.c.timestamp,
+        )
+        .label("rolling_mean_including_this_commit"),
+    ).subquery()
+
+    # Add same column, excluding the current commit
+    history = Session.query(
+        history,
+        func.coalesce(
+            func.avg(history.c.mean).over(
+                partition_by=history.c.segment_id,
+                order_by=s.extract("epoch", history.c.timestamp),
+                range_=(None, -1),
+            ),
+            history.c.mean,
+        ).label("rolling_mean_excluding_this_commit"),
+    ).subquery()
+
+    # Add column with the residuals from the exclusive rolling mean
+    history = Session.query(
+        history,
+        (history.c.mean - history.c.rolling_mean_excluding_this_commit).label(
```

Comment:
So in this case, using the mean instead of the raw data will have implications for the stats for z-score thresholds—5 sigma is not 5 sigma if the number of iterations is sometimes 1 and sometimes 100. Not sure if this is an in-practice problem, but this is worth looking into.

Reply:
Yeah at this point "z-score" is a very loose term. :) I don't think there's anything "normal" about a lot of benchmarking data